packages/core/src/auth/providers/sharedCredentialsProvider.ts

```diff
@@ -399,15 +387,51 @@ export class SharedCredentialsProvider implements CredentialsProvider {
             (k) => this.getProfile(k)
         )
 
-        return fromIni({
-            profile: this.profileName,
-            mfaCodeProvider: async (mfaSerial) => await getMfaTokenFromUser(mfaSerial, this.profileName),
-            roleAssumer: assumeRole,
-            loadedConfig: Promise.resolve({
-                credentialsFile: loadedCreds ?? profiles,
-                configFile: {},
-            } as SharedConfigFiles),
-        })
+        return async () => {
+            const iniData = loadedCreds ?? profiles
+            const profile: CredentialsData = iniData[this.profileName]
+            if (!profile) {
+                throw new ToolkitError(`auth: Profile ${this.profileName} not found`)
+            }
+            // No role to assume, return static credentials.
+            if (!profile.role_arn) {
+                return {
+                    accessKeyId: profile.aws_access_key_id!,
+                    secretAccessKey: profile.aws_secret_access_key!,
+                    sessionToken: profile.aws_session_token,
+                }
+            }
+            if (!profile.source_profile || !iniData[profile.source_profile]) {
+                throw new ToolkitError(
+                    `auth: Profile ${this.profileName} is missing source_profile for role assumption`
+                )
+            }
+            // Use source profile to assume IAM role based on role ARN provided.
+            const sourceProfile = iniData[profile.source_profile!]
+            const stsClient = new DefaultStsClient(this.getDefaultRegion() ?? 'us-east-1', {
+                accessKeyId: sourceProfile.aws_access_key_id!,
+                secretAccessKey: sourceProfile.aws_secret_access_key!,
+            })
+            // Prompt for MFA Token if needed.
+            const assumeRoleReq = profile.mfa_serial
+                ? {
+                      RoleArn: profile.role_arn,
+                      RoleSessionName: 'AssumeRoleSession',
+                      SerialNumber: profile.mfa_serial,
+                      TokenCode: await getMfaTokenFromUser(profile.mfa_serial, this.profileName),
+                  }
+                : {
+                      RoleArn: profile.role_arn,
+                      RoleSessionName: 'AssumeRoleSession',
+                  }
```

Comment:
You could technically simplify this by doing something like:
```typescript
const assumeRoleReq = {
    RoleArn: profile.role_arn,
    RoleSessionName: 'AssumeRoleSession',
    ...(profile.mfa_serial ? {
        SerialNumber: profile.mfa_serial,
        TokenCode: await getMfaTokenFromUser(profile.mfa_serial, this.profileName)
    } : {})
}
```

Reply:
Clever, I like it.